### ISSUE TITLE
Returning values from `program.action()`

### ISSUE THREAD

**Issue by acp21** [tj/commander.js#1776]
Several years ago, issue #735 was created to speak about providing a method to return a value from an action handler. Please see the following code example.

```
program
  .command('add')
  .description('Add two numbers')
  .argument('<num1>')
  .argument('<num2>')
  .action((num1, num2) =>{
    let sum = num1 + num2;
    return sum;
  });
  
let sum = program.parse()

console.log(sum)
```

Example run below:

`$ ./program add 2 5`
The anticipated output of this would be `7`.

To my knowledge, there is currently no way to return a value from an action handler like this. Is there any methodology that is recommended for achieving something like this?

**Comment by shadowspawn** [comment]
I found one other similar issue: #1503

**Comment by shadowspawn** [comment]
> To my knowledge, there is currently no way to return a value from an action handler like this.

Correct. Commander does not have a mechanism for retaining or returning a value like this.

> Is there any methodology that is recommended for achieving something like this?

Anything that retains or processes the "result" outside Commander would be fine. Like making `sum` a global variable. Or calling `console.log` inside the action handler.

I expect your real program is more challenging! 😄 Is there anything special about the pattern of your program and what you are hoping to return that might help with suggestions from readers?

**Comment by acp21** [comment]
Yes, in particular I'm trying to run commander on a server and execute different database queries based on commands sent over HTTP. Here is an example of my code below
```
program.command("count")
        .description("Count the total amount of some item.")
        .argument("<trackable>", "Define item to count.")
        .argument("<user>", "user to count trackables of")
        .action(async function (trackable) {
            console.log(`counting countable ${trackable}`);
            let count = new Count(trackable);
            ret_string = await count.run();
        });
```

and here is the server code. I'm doing this all with the `http` module as I would like to avoid installing express if possible

```
const requestListener = function (req: IncomingMessage, res: ServerResponse) {
        res.setHeader("Content-Type", "text/plain");
        // Set HTTP Code as 200 by default, only change if error occurs
        switch(req.url) {
            // Run a command on the server
            case "/command":
                res.writeHead(200);
                // Data storage string
                let command: string = '';

                // Extract data from request chunk by chunk
                req.on('data', chunk => {
                    command += chunk;
                });
                req.on('end', async () => {
                    // Log data once all received
                    let com: string = JSON.parse(command).command;
                    let arr: Array<string> = com.split(' ');
                    console.log(com)
                    let tuple: Readonly<string[]> = arr;
                    await program.parseAsync(tuple, {from: 'user', res: res})
                    res.write(ret_string);
                     res.end();
                })
```

my current goal was to use `ret_string` as a global variable that would store the value generated in the action handler callback, but this isn't ideal. Since this will be running on a server accessed by multiple users at once, this method seems like an easy way to have users receive the wrong data. Furthermore, despite the `await`s present in the server code, it does not actually wait and immediately ends the response, not giving the command enough time to run and query the DB, thusly returning an empty string.

A more ideal approach would be passing the `res` object to the command and terminating it there, as this would force it to wait for the other code to my knowledge, but so far I have been unable to find a way to pass complex data types to one of these action handler callbacks, instead only being able to use strings as if it were data being input into a CLI.

While I would prefer to avoid this, I have also considered converting the server to using the express package instead of relying on the simple http package as it may not struggle with the issues regarding immediate response sending that I have outlined above.

Please let me know if you have any insight into this issue, it has been stumping me for quite a while.

Thank you for the help!

**Comment by shadowspawn** [comment]
> Furthermore, despite the awaits present in the server code, it does not actually wait and immediately ends the response, not giving the command enough time to run and query the DB, thusly returning an empty string.

I expect that the callback for the `.on()` EventEmitter does not support async functions.

**Comment by shadowspawn** [comment]
Disclaimer: this is not a pattern I have tried in a real program! Playing with an idea to store state that supports multiple commands being run in parallel.

A way of passing in or passing out state is to store things on the command itself. That feels intrusive unless actually making a subclass of Command.

Instead, could use a Map and use the command as the key!

```js
const { program } = require('commander');

const results = new Map();
program.hook("postAction", (programHooked, actionCommand) => {
  // This is a bit subtle, but move the cache from the called command
  // to the program, where we can find it later.
  if (results.has(actionCommand)) {
    results.set(program, results.get(actionCommand));
    results.delete(actionCommand);
  }
});

program.command('foo')
  .action((options, cmd) => {
    results.set(cmd, 'Thanks for calling foo');
  });

program.command('bar')
  .action((options, cmd) => {
    results.set(cmd, 'Thanks for calling bar');
  });

program.parse();

if (results.has(program)) {
  const result = results.get(program);
  results.delete(program);
  console.log(`Retrieved result: ${result}`);
}
```

**Comment by shadowspawn** [comment]
This issue has not had any activity in a month.

Feel free to open a new issue if it comes up again, with new information and renewed interest.
 
Thank you  for your contributions.
